src/query/atomWithQuery.ts

```diff
@@ -22,83 +27,95 @@ export function atomWithQuery<
+    const state = {
+      resolve: null as ((data: TData) => void) | null,
+      setData: null as ((data: TData) => void) | null,
+      prevData: null as TData | null,
+      observer: null as QueryObserver<
+        TQueryFnData,
+        TError,
+        TData,
+        TQueryData
+      > | null,
+      handle(result: QueryObserverResult<TData, TError>) {
+        // TODO error handling
+        if (
+          result.data === undefined ||
+          (this.prevData !== null && equalityFn(this.prevData, result.data))
+        ) {
+          return
+        }
+        this.prevData = result.data
+        if (this.resolve) {
+          this.resolve(result.data)
+          this.resolve = null
+        } else if (this.setData) {
+          this.setData(result.data)
+        } else {
+          throw new Error('setting data without mount')
+        }
+      },
+    }
+    return state
+  })
+  const initAtom = atom(
     (get) => {
       const options =
         typeof createQuery === 'function' ? createQuery(get) : createQuery
-      const initAtom = atom(
-        null,
-        (get, set, cleanup: (callback: () => void) => void) => {
-          set(
-            dataAtom,
-            new Promise<TData>(() => {}) // new fetch
-          )
-          const queryClient = getQueryClient(get, set)
-          const observer = new QueryObserver(queryClient, options)
-          let hasData = false
-          observer.subscribe((result) => {
-            // TODO error handling
-            if (result.data !== undefined) {
-              if (!hasData || !equalityFn(get(dataAtom), result.data)) {
-                hasData = true
-                set(dataAtom, result.data)
-              }
-            }
-          })
-          cleanup(() => observer.destroy())
-        }
+      const state = get(stateAtom)
+      const dataAtom = atom<TData | Promise<TData>>(
+        new Promise<TData>((resolve) => {
+          state.resolve = resolve
+        })
       )
-      initAtom.onMount = (init) => {
-        let destroy: (() => void) | undefined | false
-        const cleanup = (callback: () => void) => {
-          if (destroy === false) {
-            callback()
-          } else {
-            destroy = callback
-          }
-        }
-        init(cleanup)
-        return () => {
-          if (destroy) {
-            destroy()
-          }
-          destroy = false
-        }
-      }
-      return [options, initAtom]
+      state.prevData = null
+      const queryClient = get(queryClientAtom)
+      const observer = new QueryObserver(queryClient, options)
+      state.observer?.destroy()
+      observer.subscribe((result) => {
+        state.handle(result)
+      })
+      state.observer = observer
+      return { dataAtom, options }
     },
-    (get, set, action) => {
-      if (action.type === 'refetch') {
-        const [options] = get(queryAtom)
+    (get, set, action: Action | { type: 'mount' }) => {
+      if (action.type === 'mount') {
```

Comment:
If you use a reducer type pattern, might aswell use a switch

Reply:
do you prefer switch?

Reply:
For reducer type code, I do

Reply:
0336383